### PR TITLE
storage: add TPCH load generator

### PR DESCRIPTION
**Pull request by mjibson**
Implement a TPCH load generator. It's got a few things not exactly to spec: comment text and some supplier comment exceptions. (Cockroach has the same exceptions, so perhaps not a huge deal.) On my machine, loads SF 1 in 16s with 1.86 million rows generated.

Refactor the load generator source to produce rows with an extra message if it's an end-of-txn row. Previously this was done with a vec, but because we need to produce the entire TPCH set in a single txn, we needed a way to do that without allocating them all first.

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Review by vmarcos — COMMENTED** [review]
Thanks, Matt, the ergonomics and performance of this load generator source are much improved wrt. loading via tables or other types of external sources! I tried validating the database sizes, but I am facing some troubles. I issued the following commands:

```
materialize=> CREATE SOURCE lgtpch FROM LOAD GENERATOR TPCH SCALE FACTOR 1;
CREATE SOURCE
materialize=> CREATE VIEWS FROM SOURCE lgtpch;
CREATE VIEWS
materialize=> \timing
Timing is on.
materialize=> SET transaction_isolation TO serializable;
SET
Time: 0.453 ms
materialize=> SELECT COUNT(*) FROM lineitem;
 count 
-------
     0
(1 row)

Time: 7.645 ms
materialize=> SELECT COUNT(*) FROM lineitem;
  count  
---------
 5996924
(1 row)

Time: 30572.127 ms (00:30.572)
materialize=> SELECT COUNT(*) FROM nation;
 count 
-------
    23
(1 row)

Time: 30996.546 ms (00:30.997)
```

At this point, I can see the following:

```
materialize=> select row_data from lgtpch where table = 'finished';
        row_data         
-------------------------
 {1860028,19.756075291s}
(1 row)
```

According to Tables 3 and 4 in https://www.tpc.org/tpc_documents_current_versions/pdf/tpc-h_v3.0.1.pdf, the cardinality of `lineitem` at SF 1 should be 6,001,215 and the cardinality of `nation` at all scale factors should be 25. I get the latter numbers when I generate the data via `dbgen`. But it currently does not match the output of the load generator.

The data in `nation` seems to be missing 'ALGERIA' and 'ARGENTINA'. 

To be on the safe side, perhaps some extra checks can be added to `test/testdrive/tpch.td` to verify that cardinalities and at least some of the data match what is produced by `dbgen`?

**Comment by mjibson** [comment]
Fixed the nations relation. No idea how I missed those two.

Added cardinality tests for all relations.

**Review by benesch — COMMENTED** [review]
Don’t forget to add user-facing docs!

**Comment by vmarcos** [comment]
No worries! The `nation` table looks good now, but I am still a bit worried about the cardinality of `lineitem`. Your changes to check cardinalities, including of `lineitem`, in `tpch.td` are semantically perfect according to the data generation procedure for TPC-H. That being said, the spec states specific cardinalities for `lineitem` in Table 4 of https://www.tpc.org/tpc_documents_current_versions/pdf/tpc-h_v3.0.1.pdf. I am afraid that we need to match them exactly for "all authorized scale factors"?

It looks like `dbgen` goes through quite some troubles to control random number seeding: https://github.com/electrum/tpch-dbgen/blob/32f1c1b92d1664dba542e927d23d86ffa57aa253/rnd.h#L58-L108. Also, the folks at CockroachDB are matching the exact query answers at SF 1: https://github.com/cockroachdb/cockroach/blob/master/pkg/workload/tpch/expected_rows.go. I checked manually the answers for Q1 with an interval of 90 days and they match with CockroachDB's expected rows when we run with tables loaded from `dbgen`, but do not match with data from the load generator. Perhaps we should make these match exactly to have runs that we can advertise numbers for?

**Comment by mjibson** [comment]
Cockroach doesn't match the dbgen lineitem count, theeir sf1 comes with 6002293 rows. The query answers they have their are for their own output, not dbgen's output. I believe that matching dbgen exactly is impossible unless we fully re-implement whatever rand library they are using in rust, which we aren't going to do.

```
~@work$ docker exec -it materialize-cockroach-1 cockroach workload init tpch --data-loader=import
I221003 16:22:53.622973 1 ccl/workloadccl/fixture.go:318  [-] 1  starting import of 8 tables
I221003 16:22:53.940126 129 ccl/workloadccl/fixture.go:483  [-] 2  imported 428 B in region table (5 rows, 0 index entries, took 247.928977ms, 0.00 MiB/s)
I221003 16:22:54.047343 128 ccl/workloadccl/fixture.go:483  [-] 3  imported 2.5 KiB in nation table (25 rows, 25 index entries, took 355.124533ms, 0.01 MiB/s)
I221003 16:22:54.258385 131 ccl/workloadccl/fixture.go:483  [-] 4  imported 1.6 MiB in supplier table (10000 rows, 10000 index entries, took 566.173535ms, 2.79 MiB/s)
I221003 16:22:54.782502 130 ccl/workloadccl/fixture.go:483  [-] 5  imported 26 MiB in part table (200000 rows, 0 index entries, took 1.090252254s, 23.98 MiB/s)
I221003 16:22:54.827125 133 ccl/workloadccl/fixture.go:483  [-] 6  imported 27 MiB in customer table (150000 rows, 150000 index entries, took 1.134897716s, 23.72 MiB/s)
I221003 16:22:58.884301 132 ccl/workloadccl/fixture.go:483  [-] 7  imported 130 MiB in partsupp table (800000 rows, 800000 index entries, took 5.192050625s, 25.11 MiB/s)
I221003 16:23:01.582066 134 ccl/workloadccl/fixture.go:483  [-] 8  imported 213 MiB in orders table (1500000 rows, 3000000 index entries, took 7.889840846s, 26.95 MiB/s)
I221003 16:25:50.430392 135 ccl/workloadccl/fixture.go:483  [-] 9  imported 1.5 GiB in lineitem table (6002293 rows, 48018344 index entries, took 2m56.738185507s, 8.85 MiB/s)
I221003 16:25:50.430528 1 ccl/workloadccl/fixture.go:326  [-] 10  imported 1.9 GiB bytes in 8 tables (took 2m56.800899072s, 11.09 MiB/s)
```

**Comment by mjibson** [comment]
> I checked manually the answers for Q1 with an interval of 90 days and they match with CockroachDB's expected rows when we run with tables loaded from dbgen, but do not match with data from the load generator.

Restating so I make sure I understand: you generated data with dbgen, loaded it into materialize, run q1, and the output data matches the output exactly at https://github.com/cockroachdb/cockroach/blob/b5e54ff82c685229896f9354dcee4186293cf00c/pkg/workload/tpch/expected_rows.go#L41? If so, I'm really misunderstanding something because my belief is that query output from the cockroach data can never match query output from dbgen because cockroach uses a different rng library, and also doesn't do comment generation to spec. See, for example, the comment output at https://github.com/cockroachdb/cockroach/blob/4d987b3f0788c98fa358715b336a1c55e254ee50/pkg/workload/tpch/expected_rows.go#L47 which has non-spec comments.

**Comment by vmarcos** [comment]
OK, something is going on here, since there are definitely some additional details that were not initially apparent. Here is the output we get in Materialize with running `dbgen`, loading into tables following the procedure in https://www.notion.so/materialize/Running-TPC-H-with-Tables-or-Sources-in-Materialize-205028a7d4fa45a2ba8e2823913b93d3, then running Q1:

```
 l_returnflag | l_linestatus | sum_qty  | sum_base_price  |  sum_disc_price   |     sum_charge     |      avg_qty       |     avg_price      |       avg_disc       | count_order 
--------------+--------------+----------+-----------------+-------------------+--------------------+--------------------+--------------------+----------------------+-------------
 A            | F            | 37734107 |  56586554400.73 |    53758257134.87 | 55909065222.827692 | 25.522005853257337 | 38273.129734621674 | 0.049985295838397614 |     1478493
 N            | F            |   991417 |   1487504710.38 |   1413082168.0541 |  1469649223.194375 | 25.516471920522985 |   38284.4677608483 |   0.0500934266742163 |       38854
 N            | O            | 74476040 | 111701729697.74 | 106118230307.6056 | 110367043872.49701 |  25.50222676958499 |  38249.11798890827 | 0.049996586053704085 |     2920374
 R            | F            | 37719753 |   56568041380.9 |   53741292684.604 | 55889619119.831932 |  25.50579361269077 |  38250.85462609966 |  0.05000940583012706 |     1478870
(4 rows)
```

Contrast this with the expected output of CockroachDB:

```
	1: {
		{`A`, `F`, `3.7734107e+07`, `5.65865544007299e+10`, `5.375825713486514e+10`, `5.590906522282561e+10`, `25.522005853257337`, `38273.1297346216`, `0.04998529583825443`, `1478493`},
		{`N`, `F`, `991417`, `1.4875047103799965e+09`, `1.413082168054104e+09`, `1.4696492231943603e+09`, `25.516471920522985`, `38284.467760848216`, `0.05009342667419324`, `38854`},
		{`N`, `O`, `7.447604e+07`, `1.1170172969773557e+11`, `1.0611823030761223e+11`, `1.1036704387249208e+11`, `25.50222676958499`, `38249.11798890675`, `0.04999658605362673`, `2920374`},
		{`R`, `F`, `3.7719753e+07`, `5.656804138090447e+10`, `5.374129268460378e+10`, `5.588961911982966e+10`, `25.50579361269077`, `38250.85462610268`, `0.050009405829983596`, `1478870`},
	},
``` 

The integer answers match exactly, but the fractional answers are within a 0.02 or so precision AFAICT: https://github.com/cockroachdb/cockroach/blob/34f914a90e60fc9c1d663801f0a00afa7ef51f85/pkg/workload/tpch/tpch.go#L401-L435. It is really pretty close. Q1 and Q6 are good queries to see this effect because they touch `lineitem` only.

For the first row of Q2, here are the results for Materialize over `dbgen`-generated tables vs. expected rows of CockroachDB:

```
Ours:
   9938.53 | Supplier#000005359        | UNITED KINGDOM            |    185358 | Manufacturer#4            | QKuHYh,vZGiwu2FWEJoLDx04                 | 33-429-790-6131 | uriously regular requests hag

CockroachDB expected:
		{`9938.53`, `Supplier#000005359`, `UNITED KINGDOM`, `185358`, `Manufacturer#4`, `QKuHYh,vZGiwu2FWEJoLDx04`, `33-429-790-6131`, `uriously regular requests hag`},
```

These are far too similar... For reference, with the load generator data, we get the following answers for Q1:

```
 l_returnflag | l_linestatus | sum_qty  | sum_base_price | sum_disc_price  |    sum_charge     |      avg_qty       |     avg_price      |       avg_disc       | count_order 
--------------+--------------+----------+----------------+-----------------+-------------------+--------------------+--------------------+----------------------+-------------
 A            | F            | 37750971 |    56607685743 |  54343745303.68 |  57062301169.0198 | 25.493563631648755 | 38227.669389736926 |  0.03998492035407792 |     1480804
 N            | F            |   982181 |     1471870160 |   1412849788.82 |   1483421706.6374 | 25.507219654079883 |   38224.4367111619 |  0.04004311016464967 |       38506
 N            | O            | 74303669 |   111413053734 | 106954345484.28 | 112304367432.4565 | 25.501018445969855 |  38236.95890969482 | 0.040020042879406734 |     2913753
 R            | F            | 37743107 |    56570116956 |  54308693199.38 |  57024708835.0183 | 25.509631560129577 |  38234.34146163428 | 0.039973255616692224 |     1479563
(4 rows)
```

So we are still a bit off compared to CockroachDB's expected rows and to the output we get by querying tables loaded via `dbgen`.

For Q2, I faced a schema issue so the query initially failed:

```
materialize=> SELECT
	s_acctbal,
	s_name,
	n_name,
	p_partkey,
	p_mfgr,
	s_address,
	s_phone,
	s_comment
FROM
	part,
	supplier,
	partsupp,
	nation,
	region
WHERE
	p_partkey = ps_partkey
	AND s_suppkey = ps_suppkey
	AND p_size = 15
	AND p_type LIKE '%BRASS'
	AND s_nationkey = n_nationkey
	AND n_regionkey = r_regionkey
	AND r_name = 'EUROPE'
	AND ps_supplycost = (
		SELECT
			min(ps_supplycost)
		FROM
			partsupp,
			supplier,
			nation,
			region
		WHERE
			p_partkey = ps_partkey
			AND s_suppkey = ps_suppkey
			AND s_nationkey = n_nationkey
			AND n_regionkey = r_regionkey
			AND r_name = 'EUROPE'
	)
ORDER BY
	s_acctbal DESC,
	n_name,
	s_name,
	p_partkey;
ERROR:  column reference "s_comment" is ambiguous
materialize=> \d part
                    View "public.part"
    Column     |  Type   | Collation | Nullable | Default 
---------------+---------+-----------+----------+---------
 p_partkey     | bigint  |           |          | 
 p_name        | text    |           |          | 
 p_mfgr        | text    |           |          | 
 p_brand       | text    |           |          | 
 p_type        | text    |           |          | 
 p_size        | integer |           |          | 
 p_container   | text    |           |          | 
 p_retailprice | numeric |           |          | 
 s_comment     | text    |           |          | 
```

So `s_comment` should be renamed to `p_comment` in `part`. By changing the query a bit, I then get the following first row:

```
   9990.83 | Supplier#000009280 | FRANCE         |     64267 | Manufacturer#1 | N6Gx7Dg3Y8UrZ,i3880o6xvV3cbC0uAp4Yq      | 16-470-583-1089 | zuqlCsDfz0W5BVQeBZOBTbpaxNQJGICSUh1NjgrcfeNDHvUr8FLUG56JqZYCIjCWaUUU47iko8Z2g0LVAND7tsEZXJMU8H0gDj
```

So indeed here there is a bigger difference in the query answers. The query answers published by `dbgen` are here: https://github.com/electrum/tpch-dbgen/tree/master/answers.

**Comment by mjibson** [comment]
I think that the data in https://github.com/cockroachdb/cockroach/blob/master/pkg/workload/tpch/expected_rows.go describes the dbgen expected output. When I load cockroach using `cockroach workload init tpch --data-loader=import` then run q1, I get:

```
  l_returnflag | l_linestatus |    sum_qty    |     sum_base_price     |     sum_disc_price     |       sum_charge       |      avg_qty       |     avg_price      |       avg_disc       | count_order
---------------+--------------+---------------+------------------------+------------------------+------------------------+--------------------+--------------------+----------------------+--------------
  A            | F            | 3.7787124e+07 | 5.666667070397943e+10  | 5.383709471557262e+10  | 5.5993233044912704e+10 |  25.52232140444915 |  38274.01584272584 |  0.04997078123661952 |     1480552
  N            | F            | 1.046239e+06  | 1.5715221965214233e+09 | 1.492572318339837e+09  | 1.5523919398994217e+09 | 25.461512253291474 | 38244.924594714736 | 0.050114380352532524 |       41091
  N            | O            | 7.5614018e+07 | 1.1338682729295807e+11 | 1.0771842341875688e+11 | 1.1202813600804771e+11 |  25.50041801466137 |  38239.09335603149 |  0.04997158383880077 |     2965207
  R            | F            | 3.7681836e+07 | 5.650142658769e+10     | 5.367548585331039e+10  | 5.5821749600571785e+10 | 25.501174499255914 |  38237.32842713875 |  0.05002891758747242 |     1477651
(4 rows)
```

Which is real different than the results from https://github.com/electrum/tpch-dbgen/blob/master/answers/q1.out:

```
l|l|sum_qty                                  |sum_base_price                           |sum_disc_price                           |sum_charge                               |avg_qty                                  |avg_price                                |avg_disc                                 |count_order           
A|F|37734107.00|56586554400.73|53758257134.87|55909065222.83|25.52|38273.13|0.05|               1478493
N|F|991417.00|1487504710.38|1413082168.05|1469649223.19|25.52|38284.47|0.05|                 38854
N|O|74476040.00|111701729697.74|106118230307.61|110367043872.50|25.50|38249.12|0.05|               2920374
R|F|37719753.00|56568041380.90|53741292684.60|55889619119.83|25.51|38250.85|0.05|               1478870
```

My guess is their expected_rows.go is used to verify cockroach's correctness against dbgen, not their own workload generator. We should perhaps take the same approach: use dbgen-generated data to verify we output numbers that are within the defined correctness range, but we also have a load generator that has no guaranteed output, it's there only to make loading tpch-shaped data much easier for optimization and load testing work, not correctness.

I fixed the comment name and have rebased on main with the recent subsources changes.

**Comment by vmarcos** [comment]
Cool, thanks, Matt, for diving down into CockroachDB's behavior. This gives us a clear point of comparison and I agree with you that we can similarly scope the TPC-H load generator to loading data and investigating performance effects, but use `dbgen` when we need "official" numbers and a guarantee of result correctness according to the standard. Perhaps it might be worth adding a small remark to the user-facing docs that we implement the TPC-H benchmark specification, modulo differences in random number generation wrt. `dbgen`?

As a final validation step, I was going to execute all the TPC-H queries to contrast the performance with 1 worker using data from the load generator vs. `dbgen` at SF 1. However, I am now facing some troubles running the code with `bin/environmentd --release --reset`. After I issue: 

````
materialize=> CREATE SOURCE lgtpch FROM LOAD GENERATOR TPCH (SCALE FACTOR 1) FOR ALL TABLES;
CREATE SOURCE
Time: 1675.947 ms (00:01.676)
````

I am getting the following repeatedly in the logs:

````
environmentd: 2022-10-10T09:42:11.809440Z  INFO mz_orchestrator_process: Launching storage-u9: --workers=1 --controller-listen-addr=127.0.0.1:2103 --internal-http-listen-addr=127.0.0.1:2104 --opentelemetry-resource=storage_id=u9 --log-filter=info --log-prefix=storage-u9 --pid-file-location=/Users/vmarcos/workspace/temp/materialize/mzdata/storage-u9-0.pid --secrets-reader=process --secrets-reader-process-dir=/Users/vmarcos/workspace/temp/materialize/mzdata/secrets...
storage-u9: 2022-10-10T09:42:11.827572Z  INFO storaged: about to bind to 127.0.0.1:2103
storage-u9: 2022-10-10T09:42:11.827609Z  INFO storaged: serving internal HTTP server on 127.0.0.1:2104
storage-u9: 2022-10-10T09:42:11.829027Z  INFO mz_service::grpc: Starting to listen on 127.0.0.1:2103
storage-u9: 2022-10-10T09:42:14.698535Z  INFO mz_service::grpc: GrpcServer: remote client connected
environmentd: 2022-10-10T09:42:14.699070Z  INFO mz_service::grpc: GrpcClient localhost:2103: connected
storage-u9: 2022-10-10T09:42:16.593794Z  INFO tokio_postgres::connection: NOTICE: relation "consensus" already exists, skipping    
thread 'timely:work-0' panicked at 'called `Option::unwrap()` on a `None` value', src/storage/src/source/source_reader_pipeline.rs:887:75
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:48:5
   3: core::option::Option<T>::unwrap
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/option.rs:775:21
   4: mz_storage::source::source_reader_pipeline::remap_operator::{{closure}}::{{closure}}
             at ./src/storage/src/source/source_reader_pipeline.rs:887:32
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/mod.rs:91:19
   6: <futures_util::future::future::fuse::Fuse<Fut> as core::future::future::Future>::poll
             at /Users/vmarcos/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.24/src/future/future/fuse.rs:86:37
   7: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/future.rs:124:9
   8: <timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G> as mz_timely_util::operators_async_ext::OperatorBuilderExt<G>>::build_async::{{closure}}::{{closure}}
             at ./src/timely-util/src/operators_async_ext.rs:165:43
   9: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/dataflow/operators/generic/builder_rc.rs:162:26
  10: <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/dataflow/operators/generic/builder_raw.rs:204:9
  11: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:684:30
  12: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:344:26
  13: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:310:17
  14: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:684:30
  15: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:344:26
  16: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/progress/subgraph.rs:310:17
  17: timely::worker::Wrapper::step::{{closure}}
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/worker.rs:754:57
  18: core::option::Option<T>::map
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/option.rs:929:29
  19: timely::worker::Wrapper::step
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/worker.rs:754:26
  20: timely::worker::Worker<A>::step_or_park
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/worker.rs:396:38
  21: mz_storage::storage_state::Worker<A>::run_client
             at ./src/storage/src/storage_state.rs:134:17
  22: mz_storage::storage_state::Worker<A>::run
             at ./src/storage/src/storage_state.rs:110:33
  23: mz_storage::protocol::server::serve::{{closure}}
             at ./src/storage/src/protocol/server.rs:94:9
  24: timely::execute::execute::{{closure}}
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/timely/src/execute.rs:287:22
  25: timely_communication::initialize::initialize_from::{{closure}}
             at /Users/vmarcos/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/9548bf9/communication/src/initialize.rs:316:33
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
environmentd: 2022-10-10T09:42:25.865863Z ERROR mz_orchestrator_process: storage-u9 exited: Ok(ExitStatus(unix_wait_status(6))); relaunching in 5s
environmentd: 2022-10-10T09:42:25.865892Z  WARN mz_storage::controller::rehydration: storage host produced error, reconnecting: status: Unknown, message: "error reading a body from connection: broken pipe", details: [], metadata: MetadataMap { headers: {} }
````

Not sure whether this is independent of the load generator, but just happens to be triggered by it after the rebase on main? The test at scale factor .001 works, so it appears that we need to generate more data to trigger the behavior above.

**Comment by mjibson** [comment]
That crash is happening due to a race condition bug in storage. I've asked them to take over fixing it. I'm also going to wait on merging this until #15203 is merged, otherwise the entire TPCH dataset is stored in memory until it's all ready, which we also don't want.

**Comment by mjibson** [comment]
Those bugs have been fixed. Rebased on main. Should be usable with any SF and mergeable now.

**Comment by vmarcos** [comment]
There is one small detail you might wish to look into for the tests. The incantation that worked for `CREATE SOURCE` for me required a `SIZE` parameter to be specified. For example:

```
CREATE SOURCE lgtpch FROM LOAD GENERATOR TPCH (SCALE FACTOR 1) FOR ALL TABLES WITH (SIZE = '4');
```

I then went ahead and executed all of the TPC-H queries with PK and FK indexes locally (similar setup as in this [doc](https://www.notion.so/materialize/TPC-H-Optimization-and-Dataflow-Performance-d43aa9cb379d46ccaa0be07e83d658db)). I tried SF 1 with 1 and 8 workers as well as SF 4 with 4 workers. I summarized the results in this [spreadsheet](https://docs.google.com/spreadsheets/d/1w-Q40pOkWT568wmtONTgJAU0Yk-PodF1gGtFes6H72Q/edit#gid=0).

Most queries exhibit the same performance effects with data generated by  `dbgen` or by the load generator, but there are some differences. I marked in red the more concerning queries, IMHO.

I tried to perform some basic validation at scale factor 1 with 1 worker. Q5 has the same number of rows in the result with both `dbgen` and load generator data, but exhibits minor plan differences. Q7 has the same plan and the same number of rows in the result. Q21 has the same plan and a similar number of rows in the result (411 vs. 405, respectively). Especially with Q7 and Q21, it is not straightforward to see why the query processing performance differs so much. I did not dig down into physical plans, but that would be the next step.

I think tracking down these differences exceeds the scope of this PR. I would suggest to go ahead with merging (after fixing the minor test issues), since the load generator will already be extremely useful and yield much higher productivity than following the procedure with `dbgen` and tables. The load generator is _so_ much faster and convenient! Then we could create a separate issue to look into the performance differences, probably to be allocated to the COMPUTE team and further prioritized.

What do you think @mjibson and @benesch? Should I go ahead with creating the separate issue for COMPUTE to look into query processing time differences?

**Review by vmarcos — APPROVED** [review]

**Comment by benesch** [comment]
> Should I go ahead with creating the separate issue for COMPUTE to look into query processing time differences?

SGTM!

**Pull request merged**
